### PR TITLE
UI/code tab

### PR DESCRIPTION
**Pull request by odedre**
## Proposed Changes

- added code tab
- added code aspect
- added queries to get the files, dependencies and code data for a component

TODO
- [x] mdx highlight
- [x] code snippet empty state

### CODE REVIEW

File: scopes/ui-foundation/tree/file-tree/file-tree.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  onSelect?: (id: string, event?: React.MouseEvent) => void;
+  selected?: string;
+  files: string[];
+  TreeNode?: any; //TreeNodeRenderer<PayloadType>; - what do we want here?
+};
+
+export function FileTree({ files, onSelect, selected, TreeNode }: FileTreeProps) {
+  const rootNode = useMemo(() => {
+    const tree = inflateToTree(files, (c) => c);
+    return tree;
```
Comment (KutnerUri) [inline scopes/ui-foundation/tree/file-tree/file-tree.tsx:18]: for dev/main labels, choose one of these solutions:

* calc in widget (node) => devFiles.has(node.id)
* attachPayload (map<Node, payload>() )
* files.map((filepath) => { id: filepath, isDev: true, ... }

File: scopes/ui-foundation/tree/tree-node/tree-node.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      isActive={() => isActive}
+      exact
+      strict
+      className={classNames(indentClass, clickable, styles.fileNode)}
+      activeClassName={styles.active}
+      onClick={onClick}
+    >
+      <div className={styles.left}>
+        <img className={styles.icon} src={icon} />
+        <span>{node.id.split('/').pop()}</span>
```
Comment (KutnerUri) [inline scopes/ui-foundation/tree/tree-node/tree-node.tsx:34]: consider using path / path-browserify

**Review by KutnerUri — CHANGES_REQUESTED** [review]

File: scopes/component/code/ui/dependency-tree/dependency-tree.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        className={styles.dependencyDrawer}
+      />
+    </div>
+  );
+}
+
+function DependencyList(deps) {
+  if (!deps) return <div />;
+  return deps.map((dep) => {
+    const linkPrefix = dep.__typename === 'ComponentDependency' ? 'https://bit.dev/' : 'https://npmjs.com/';
```
[the remark this answers is not in the document]
Comment (odedre) [reply scopes/component/code/ui/dependency-tree/dependency-tree.tsx:45]: works now. extracted to function

File: scopes/ui-foundation/tree/drawer/drawer.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import { Icon } from '@teambit/evangelist.elements.icon';
+import classNames from 'classnames';
+import React from 'react';
+
+import { DrawerType } from './drawer-type';
+import styles from './drawer.module.scss';
+
+export type DrawerProps = {
+  drawer: DrawerType;
```
[the remark this answers is not in the document]
Comment (odedre) [reply scopes/ui-foundation/tree/drawer/drawer.tsx]: done

File: scopes/ui-foundation/tree/drawer/drawer.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  isOpen: boolean;
+  onToggle: (event: React.MouseEvent<HTMLDivElement>) => void;
+} & React.HTMLAttributes<HTMLDivElement>;
+
+export function DrawerUI({ drawer, className, isOpen, onToggle, ...rest }: DrawerProps) {
```
[the remark this answers is not in the document]
Comment (odedre) [reply scopes/ui-foundation/tree/drawer/drawer.tsx]: done

File: scopes/component/code/ui/code-tab-tree/code-tab-tree.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      onToggleDrawer((list) => list.filter((drawer) => drawer !== id));
+      return;
+    }
+    onToggleDrawer((list) => list.concat(id));
+  };
+
+  // TODO - handle labels for main file or dev
+  const widgets = [];
+
+  const TreeNodeRenderer = useCallback(
```
[the remark this answers is not in the document]
Comment (odedre) [reply scopes/component/code/ui/code-tab-tree/code-tab-tree.tsx:36]: passed widgets but left out of file tree

File: scopes/ui-foundation/tree/file-tree/file-tree.tsx (hunk 1 above)
Comment (odedre) [reply scopes/ui-foundation/tree/file-tree/file-tree.tsx:18]: pass from code tree

File: scopes/component/code/ui/code-view/code-view.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  const title = useMemo(() => currentFile?.split('/').pop(), [currentFile]);
+  const lang = useMemo(() => {
+    const langFromFileEnding = currentFile?.split('.').pop();
+
+    // for some reason, SyntaxHighlighter doesnt support scss or sass highlighting, only css. I need to check how to fix this properly
+    if (langFromFileEnding === 'scss' || langFromFileEnding === 'sass') return 'css';
+    return langFromFileEnding;
+  }, [fileContent]);
+
+  if (!fileContent) return null; // is there a state where the is no file content? what should be presented then?
```
Comment (KutnerUri) [inline scopes/component/code/ui/code-view/code-view.tsx:26]: maybe show empty state?

File: scopes/component/code/ui/dependency-tree/dependency-tree.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+          <span>{`${dependency.name}@${dep.version}`}</span>
+        </Link>
+      </div>
+    );
+  });
+}
+
+// remove this once the links are calculated in the dependency resolver
+function getDependencyLink(dep: DependencyType) {
+  const version = dep.version.replace('^', '');
```
Comment (KutnerUri) [inline scopes/component/code/ui/dependency-tree/dependency-tree.tsx:54]: add '~'

File: scopes/ui-foundation/get-icon-from-file-name/get-icon-from-file-name.tsx [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import { getIconForFile } from 'vscode-icons-js';
+
+export function getIcon(fileName?: string) {
+  if (!fileName) return '';
+  const iconName = getIconForFile(fileName);
+  const storageLink = 'https://static.bit.dev/file-icons/';
```
Comment (KutnerUri) [inline scopes/ui-foundation/get-icon-from-file-name/get-icon-from-file-name.tsx:6]: use storage link

**Review by KutnerUri — APPROVED** [review]

File: scopes/code/queries/get-component-code/get-component-code.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      mainFile?: string;
+      dependencies: DependencyType[];
+      devFiles: string[];
+    };
+  };
+};
+
+export function useCode(componentId: ComponentID) {
+  const id = componentId._legacy.name;
+  const { data } = useDataQuery<CodeResults>(getCode, {
```
[the remark this answers is not in the document]
Comment (odedre) [reply scopes/code/queries/get-component-code/get-component-code.ts]: done

File: scopes/component/code/code.ui.runtime.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+  static slots = [Slot.withType<string>()];
+
+  static async provider([component]: [ComponentUI], config, [fileIconSlot]: [FileIconSlot]) {
+    const ui = new CodeUI(fileIconSlot);
+    const section = new CodeSection(ui);
+
+    // overrides the default tsx react icon with the typescript icon
+    ui.registerEnvFileIcon([
+      (fileName) => (isTsx.test(fileName) ? 'https://static.bit.dev/file-icons/file_type_typescript.svg' : undefined),
```
[the remark this answers is not in the document]
Comment (odedre) [reply scopes/component/code/code.ui.runtime.tsx]: done

File: scopes/component/code/ui/dependency-tree/dependency-tree.tsx (hunk 8 above)
Comment (odedre) [reply scopes/component/code/ui/dependency-tree/dependency-tree.tsx:54]: done

File: scopes/ui-foundation/get-icon-from-file-name/get-icon-from-file-name.tsx (hunk 9 above)
Comment (odedre) [reply scopes/ui-foundation/get-icon-from-file-name/get-icon-from-file-name.tsx:6]: done

File: scopes/component/code/ui/code-view/code-view.tsx (hunk 7 above)
Comment (odedre) [reply scopes/component/code/ui/code-view/code-view.tsx]: done

**Pull request merged**